omod/src/main/java/org/openmrs/web/attribute/handler/SerializingFieldGenDatatypeHandler.java

```diff
@@ -33,6 +44,37 @@ public T getValue(DT datatype, HttpServletRequest request, String formFieldName)
 		if (StringUtils.isBlank(formFieldValue)) {
 			return null;
 		}
+		if (isInteger(formFieldValue)) {
+			BaseOpenmrsObject metadata = null;
+			PropertyEditor editor;
+
+			if (datatype instanceof LocationDatatype) {
+				editor = new LocationEditor();
+				editor.setAsText(formFieldValue);
+				metadata = (BaseOpenmrsObject) editor.getValue();
+			} else if (datatype instanceof ProgramDatatype) {
+				editor = new ProgramEditor();
+				editor.setAsText(formFieldValue);
+				metadata = (BaseOpenmrsObject) editor.getValue();
+			} else if (datatype instanceof ConceptDatatype) {
+				editor = new ConceptEditor();
+				editor.setAsText(formFieldValue);
+				metadata = (BaseOpenmrsObject) editor.getValue();
+			} else if (datatype instanceof ProviderDatatype) {
+				editor = new ProviderEditor();
+				editor.setAsText(formFieldValue);
+				metadata = (BaseOpenmrsObject) editor.getValue();
+			}
+
+			if (metadata != null) {
+				formFieldValue = metadata.getUuid();
+			}
+		}
+
 		return datatype.deserialize(formFieldValue);
 	}
+
+	private boolean isInteger(String value) {
+		return value.matches("-?\\d+");
```

Comment:
There should be a java library for this.

Reply:
StringUtils.isNumeric() would be ok?